### PR TITLE
Allow purchase of crypto assets via Coinbase Pay #5198

### LINKED ISSUE #5198: Allow purchase of crypto assets via Coinbase Pay
Allow users to purchase crypto assets via Coinbase Pay.

--

Endpoint: 
`https://pay.coinbase.com/buy/select-asset`

Parameters:
`appId` : `string` - Provided by Coinbase Pay team, check discord for valid AlphaWallet app ID.
`destinationWallets`  : `array` - See below.

Destination Wallets Parameters:
`address` : `string` -  Destination address where the purchased tokens will be sent.
`blockchains` : `string array` - List of enabled blockchains, available values*: `ethereum`, `solana`, `avalanche-c-chain` 
`assets` : `string array` - List of tokens that will be displayed to the user in the succeeding screen

* Seems only `ethereum` blockchain is relevant to AlphaWallet

Upon testing, there are two ways we could utilize this API.
1. Select from a list of tokens in the `ethereum` network:

```
https://pay.coinbase.com/buy/select-asset?appId=<APP_ID>&destinationWallets=[{"address": "<WALLET_ADDRESS>","blockchains":["ethereum"]}
```

2. Directly purchase a token; this will skip the token selection screen:

```
https://pay.coinbase.com/buy/select-asset?appId=<APP_ID>&destinationWallets=[{"address": "<WALLET_ADDRESS>","assets":["ETH"]}
```

Requirements to test this feature:
1. Verified Coinbase account - may include ID verification and linked payment info depending on the country of origin.

--

Sample flow:

https://user-images.githubusercontent.com/17334718/185843826-c267b688-f2c9-416b-b161-1bcc87d38522.mp4

-- 

Relevant info:
* https://www.coinbase.com/pay-sdk
* https://docs.cloud.coinbase.com/pay-sdk/docs/generating-url
* https://github.com/coinbase/cbpay-js

### PR DESCRIPTION
**Pull request by oa-s**
Closes #5198

### CODE REVIEW

**Review by hboon — COMMENTED** [review]
Does it work for you? I get this after I login to Coinbase:

Screenshot 2022-08-24 at 3 58 43 PM

**Comment by hboon** [comment]
(I did set the app ID 3c854xxxxx0753)

File: AlphaWallet/Core/BuyToken/CoinBase/CoinBase.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        guard let platform = token.server.coinBasePlatform else { return nil }
+        return Constants.buyWitCoinBaseUrl(blockchain: platform, wallet: wallet).flatMap { URL(string: $0) }
+    }
+
+    func actions(token: TokenActionsIdentifiable) -> [TokenInstanceAction] {
+        return [.init(type: .buy(service: self))]
+    }
+
+    func isSupport(token: TokenActionsIdentifiable) -> Bool {
+        return token.server.coinBasePlatform != nil
```
Comment (hboon) [inline AlphaWallet/Core/BuyToken/CoinBase/CoinBase.swift:32]: Maybe should just check server == .main instead?

Comment (oa-s) [reply AlphaWallet/Core/BuyToken/CoinBase/CoinBase.swift:32]: documentation says Avalanche is supported

Comment (oa-s) [reply AlphaWallet/Core/BuyToken/CoinBase/CoinBase.swift:32]: lets make it only for main for now

**Comment by oa-s** [comment]
@hboon how about grouping swap actions in same way buy does, some of tokens support for different swap types, so we could update action sheet on token screen with single action `swap`, tapping this action will present action seet of available swap types

**Comment by oa-s** [comment]
> Does it work for you? I get this after I login to Coinbase:
> 
> Screenshot 2022-08-24 at 3 58 43 PM

i got unverified account, not able to test next steps

**Comment by hboon** [comment]
> @hboon how about grouping swap actions in same way buy does, some of tokens support for different swap types, so we could update action sheet on token screen with single action `swap`, tapping this action will present action seet of available swap types

Grouping in the UI or in the code?

**Comment by hboon** [comment]
> i got unverified account, not able to test next steps

OK. I'll test it out on my end then.

**Comment by oa-s** [comment]
something like this 
Screenshot 2022-08-24 at 11 19 35

**Comment by hboon** [comment]
> grouping swap actions in same way buy does, some of tokens support for different swap types,
> something like this

Sure!

File: AlphaWallet/Settings/Types/Constants.swift [hunk 2]
```diff
@@ -104,6 +104,11 @@ public struct Constants {
         "https://buy.ramp.network/?hostApiKey=\(Constants.Credentials.rampApiKey)&hostLogoUrl=https%3A%2F%2Falphawallet.com%2Fwp-content%2Fthemes%2Falphawallet%2Fimg%2Falphawallet-logo.svg&hostAppName=AlphaWallet&swapAsset=\(asset)"
     }
 
+    static func buyWitCoinBaseUrl(blockchain: String, wallet: Wallet) -> String? {
```
Comment (hboon) [inline AlphaWallet/Settings/Types/Constants.swift:107]: Typo. "Wit" -> "With"

**Review by hboon — APPROVED** [review]
I pushed 2 commits, 1 of them fixing the URL. So this should be good to go.

**Comment by hboon** [comment]
@oa-s would you help squash to remove my commits before merging, if they look alright?

File: AlphaWallet/BuyCrypto/BuyCryptoUsingThirtPartyCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
+//
+//  BuyCryptoUsingThirtPartyCoordinator.swift
+//  AlphaWallet
+//
+//  Created by Vladyslav Shepitko on 24.08.2022.
+//
+
+import UIKit
+
+protocol BuyCryptoUsingThirtPartyCoordinatorDelegate: AnyObject, CanOpenURL {
```
Comment (hboon) [inline AlphaWallet/BuyCrypto/BuyCryptoUsingThirtPartyCoordinator.swift:10]: "Thirt" -> "Third" :)

File: AlphaWallet/BuyCrypto/BuyCryptoUsingThirtPartyCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+class BuyCryptoUsingThirtPartyCoordinator: Coordinator {
```
Comment (hboon) [inline AlphaWallet/BuyCrypto/BuyCryptoUsingThirtPartyCoordinator.swift:13]: "Thirt" -> "Third" :)

File: AlphaWallet/Core/BuyToken/BuyTokenService.swift [hunk 5]
```diff
@@ -6,7 +6,51 @@
...
+    private let subProviders: [BuyTokenURLProviderType & SupportedTokenActionsProvider]
+
+    var objectWillChange: AnyPublisher<Void, Never> {
+        return Just(subProviders)
+            .flatMap { Publishers.MergeMany($0.map { $0.objectWillChange }) }
+            .mapToVoid()
+            .eraseToAnyPublisher()
+    }
+    let analyticsNavigation: Analytics.Navigation = .fallback
+    let analyticsName: String = "Buy Crypto"
```
Comment (hboon) [inline AlphaWallet/Core/BuyToken/BuyTokenService.swift:29]: Might be unused?

Comment (oa-s) [reply AlphaWallet/Core/BuyToken/BuyTokenService.swift:29]: yes, its looks unused but its required by protocol

Comment (oa-s) [reply AlphaWallet/Core/BuyToken/BuyTokenService.swift:29]: yes, its looks unused but its required by protocol

**Pull request merged**